### PR TITLE
Add new sanitise filter on all raw instances

### LINKED ISSUE #710: Make raw filter opt-in?
We use the `raw` filter in various places where we need to nest macros, for example:

```
{{
    html.button({
        'label': html.icon('check') ~ ' Call to Action'
    })
}}

<button class="btn btn--primary">
    <i aria-hidden="true" class="icon-ok"></i> Call to Action
</button>
```

To achieve this, we can't do any form of escaping on the `label` option and must implicitly trust that the value supplied will be safe. However, there are situations where we need to give Admins the ability to supply custom values to buttons. labels etc, so we need to come up with an alternative to using `raw` that still allows us to nest macros in the way we need. (The button example here is a simple one, there are more complex uses of `raw` in form helpers).

Without `raw`, the macro renders:

```
<button class="btn btn--primary">
    &lt;i aria-hidden="true"  class="icon-ok"&gt;&lt;/i&gt; Call to Action
</button>
```

### Option: Make `raw` opt-in behaviour

DG suggested this, and the more I try and avoid it the more I think it's going to be the only way forward. We would add a new option to helpers like button, which would allow a developer to turn on `raw` mode, it would then, depending on the design of the macro, allow unfiltered markup into the view to achieve whatever if required at the time.

```
{{
    html.button({
        'label': html.icon('check') ~ ' Call to Action',
        'raw': true
    })
}}
```

and inside the button macro we'd do something like:

```
<button{{
    attributes(options
        |exclude('caret label')
        |defaults({
            'class': 'btn'
        })
    )
}}>
    {% set label = options.label|default|strip_tags %}
    {% if options.raw is defined and options.raw != false %}
        {{- label|raw -}}
    {% else %}
        {{- label -}}
    {% endif %}
    {{- util.caret(options.caret|default) -}}
</button>
```

😎 Fail secure
😎 Make devs explicitly choose when to use raw (i.e. not when expecting user supplied inputs)
😎 Easy to search/audit where the raw flag is used in product views to sanity check the use
💀 Most impact to implement in products, would need to be part of a major version release
💀 All existing Pulsar views would need to be updated and tested

I'm not really considering suggesting it should be opt-out at this point, I can't make a good enough argument for it.

----

I'm open to any other options or new ideas, I've looked into using `string_from_template()` instead of `raw` but I don't think it would actually fix anything apart from looking less scary by not saying 'raw'. I can't create a custom filter because everything's escaped unless you use `raw` to mark it safe.

This doesn't cover all uses of `raw` either, some at the template level (like the tab helpers) simply can't be avoided, but unless we're allowing users control over that level of the template (which we don't), we should be ok.

As an aside, it's worth mentioning that this new constraint on raw would probably make us design other helpers for specific use cases, like icon buttons, which move us away from nesting the icon helper, but as I mentioned earlier form helpers are a bit of a different case.

On the plus side, we have a major version slated for early next year which will involve some end to end view testing in products, so perhaps it's an ideal time to work something like this in?

### HTML
- [x] badge
- [x] button
- [ ] countdown
- [x] div
- [x] icon
- [x] label
- [x] link
- [x] li
- [x] media
- [x] metadata
- [x] panel
- [x] progress
- [x] dropzone

### Form
- [x] content (via elem.paragraph)
- [x] end
- [x] error
- [x] fieldset_start
- [ ] group
- [x] help

### Elem
- [x] paragraph

### Nav
- [x] primary
- [x] nav_item

### Flash
- [x] message

### Rules
- [ ] block

### Tabs
- [x] content (Can't see that we can get away from using this, anything passed to tabs.content should be considered safe at the point of render)

### PR DESCRIPTION
**Pull request by Stanton**
Fixes #710

This change adds the lightweight [htmLawed](http://www.bioinformatics.org/phplabware/internal_utilities/htmLawed/rsnake/RSnakeXSSTest.htm) utility which can be used to filter HTML against all documented XSS exploits on the Snake and [OWASP XSS filter evasion cheatsheet](https://owasp.org/www-community/xss-filter-evasion-cheatsheet).

The implementation for Pulsar is a new Twig extension which provides a simplefilter which should be used in any/all places where the `raw` filter is used. This will sanitise any malicious scripts, attributes or other attempts to exploit anywhere we may allow users to pass currently unfiltered input to the view.

Initial tests using the Pulsar Lexicon don't show any visual differences using the [configuration currently defined](https://github.com/jadu/pulsar/compare/xss-remediation?expand=1#diff-738f5df8768296e284707353421232e0cd9b650a8862a4ada7b40a64d466b5dc), which is a slightly customised use of htmLawed's recommended `'safe'=>1` config for XSS remediation (It's custom so we can still allow HTML comments to be treated as such and not converted to plain text). There may be other decisions we make to refine the configuration, such as not allowing the inline `style` attribute which can be a clickjacking vector. However that's a more controversial change at this point.

## Talking points

At the moment the configuration removes all `on*` attributes, e.g. `onclick`, `onfocus`, `onblur`, `onload` will all be stripped because (in this PR) `sanitise` is being invoked when outputting tab content, this represents the largest potential blocker for products using inline attributes (or third-party code using inline attributes), there's a wider discussion to have at product level about whether we should refector out our use of `on*` attributes to more structured scripts to reduce the attack surface they represent, or whether we'll need to relax the sanitisation in certain areas like tab content, and make a bigger assumption that the content of an entire tab is deemed to be safe at the point of render.

Put simply, we have these options:
1. Continue with _strict_ sanitisation, and refactor places where we use `on*` attributes (fire, screams, sirens)
1. Don't sanitise instances of `raw` which output the larger portions of layout, only sanitise the helpers (birds, a light breeze, pleasant sunshine, with a large level of trust required that we've sanitised everything enough until this point)

## Pro/Cons of the sanitise method over raw opt-in

😃 No need to rewrite helpers to permit raw opt-in
😃 No need for product teams to update helper instances for an opt-in
😃 Relatively easy find & replace operation for product teams to replace their own instances of `|raw` with `|sanitise|raw`
😃 More performant that HTMLTidy

😐 Presumes htmLawed stays maintained and updated with new vectors, however the actual core library used by our `sanitise` filter can be changed at any time, the filter provides a level of abstraction within the view allowing us to change or update easily. That being said it's still a far smarter decision than rolling our own
😐 If products have any tests that expect markup patterns that will be documented by the _Markup changes_ section, these may now fail

💀 Product teams still need to find and replace all instances where they're using `raw`
💀 Will still need smoke testing to find any unexpected results of filtering (candidate for automated visual regression testing?)

## Markup changes

HTMLawed has a couple of opinionated quirks which change our markup in the following ways:
* Boolean attributes are altered to include the lowercase canonical name for the attribute, changing `disabled` into `disabled="disabled"` for example. ([perfectly valid though, see the HTML spec](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes)), for Pulsar this mainly affects the `checked`, `selected`, `multiple`, `disabled` and `required` attributes
* Textarea elements are provided with default `row` and `cols` attributes when omitted, Pulsar helpers have been updated to provide sensible default values in-line with expected visuals, they are still able to be overridden by Pulsar width modifiers

## Example from the button helper:

Given the following exploits:

```twig
{{ html.button({ 
  label: html.icon('ok') ~ "<script>alert('There’s some holes in this house')</script>",
  type: 'submit'
}) }}

{{ html.button({ 
  label: html.icon('ok') ~ "<strong onclick='alert(2)'>foo</strong>",
  type: 'submit'
}) }}
```

The following helper, using the new `sanitise` filter before passing to `raw`...

```twig
<button{{
  attributes(options
    |exclude('label')
    |defaults({
      'class': 'btn'
    })
  )
}}>
  {{- options.label|default|sanitise|raw -}}
</button>
```

Creates the following output, note the icons and strong text are still rendered as desired, you'll have to take my word for it that the alerts aren't fired.

Screenshot 2020-10-16 at 10 56 01

## Breaking change checklist

The following points should be considered as an early-warning of introducing a breaking change to a Continuum platform product. This is not an exhaustive list of what constitutes a breaking change.

| Does this pull request... | ⚠️ Yes / ✅ No |
| ---- | ---- |
| require changes to `main.js` | ✅ No |
| require changes to `index.js` | ✅ No |
| require changes to `pulsar.scss` | ✅ No |
| require changes to `gruntfile.js` | ✅ No |
| require changes to `package.json` (not including dev dependencies) | ✅ No |
| require changes to `base.html.twig` (or other core views like header/footer) | ✅ No |
| require changes to Twig configuration | ⚠️ Yes - The new sanitise function will need registering with other Pulsar Twig extensions, see app.php for our simplified example |
| require new arguments to be passed to a js component | ✅ No |
| require markup changes to maintain functionality | ✅ No - Some markup output will change, but it's a result of the tool being used rather than being required |
| require changes to existing unit tests | ⚠️ Yes - Some attribute formatting has changed, see 'markup changes' above |

### CODE REVIEW

**Comment by Stanton** [comment]
If this PR gets approved the AttributeParserExtension will need updating to transpose boolean attributes to use the lowercase canonical name

**Review by jackbentley — CHANGES_REQUESTED** [review]
I've got multiple concerns with this PR. It doesn't seem like it's fixing a problem to me. I can foresee it actually causing more problems than it's trying to solve.

If this PR were to be merged, it then becomes Pulsar who decides how data is sanitised and outputted. The control of the sanitisation of data is no longer the responsibility of the application using Pulsar, rather Pulsar it's self. This is a bad thing.

It's also worth noting that sanitisation of data and escaping of data are separate things. And this PR seems to be us trying to use sanitisation where we should be using escaping. Sanitisation normally happens on input, and escaping normally happens on output.

Twig is by default an [auto-escaping](https://twig.symfony.com/doc/1.x/tags/autoescape.html) templating language. This means that by default all variables are escaped unless marked as "safe" (per variable or per render). You are expected to mark variables individually as "safe" using the `|raw` filter. [The docs](https://twig.symfony.com/doc/1.x/filters/raw.html) describe this quite well.

There are some slight caveats to this, given that Twig needs to know how it's escaping the variable. By default, it's set to `HTML` escaping. If you're attempting to output a variable into anything other than an HTML context, then you need to tell twig the context. This is described in [the escape filter docs](https://twig.symfony.com/doc/1.x/filters/escape.html).

Escaping context becomes particularly important when dealing with things like JSON. See XFPS-320.

Pulsar seems to use variables in Twig in essentially two ways: To be structural (provide HTML), or, to output data.

When outputting a variable containing data, you shouldn't use the `|raw` filter as it needs to be escaped in some way. The variables are not "safe". When a variable is "structural", you should use `|raw` to output the raw HTML. We know these will be safe as they will contain hand-written code.

Sidenote: However to be entirely honest, in an ideal world, there shouldn't really be many if any places that use "structural" variables. They should actually be an [`embed`](https://twig.symfony.com/doc/1.x/tags/embed.html) and not an include with variables passed to it. 

With this change, certain HTML in those "structural" variables would end up being stripped by this `|sanitise` filter.

Additionally, what this is achieving isn't what I'd expect when I pass user data to a helper. If I was passing a variable containing the user data `<strong onclick='alert(2)'>foo</strong>` then I would expect exactly that variable to be outputted as text. As it's user data, the HTML should not be respected at all and should be escaped entirely. And the value of the content should not change at all. This is crucial for things like `form.content()`.

**TL;DR:**

Blindly adding a `sanitise` filter will cause no end of problems. As mentioned, this likely isn't what we want when outputting user data. We probably want to be encoding. The correct solution would be to use the `raw` and `encode` filters already in Twig more appropriately for the context. We need to go through and understand every point that is using `|raw` and decide if it should or shouldn't.

File: views/pulsar/v2/helpers/elem.html.twig [hunk 1]
```diff
@@ -415,7 +418,9 @@ value       | string  | Specifies the value of the input
         attributes(options
             |exclude('append bare has_error help_id prepend value')
             |defaults({
-                'class': 'form__control'
+                'class': 'form__control',
+                'cols': 20,
+                'rows': 2
```
[the remark this answers is not in the document]
Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig:423]: Documented in the PR, Pulsar's width modifiers still work as expected, and the `20` cols is very close to the default width when no modifier is provided.

File: tests/unit/Jadu/Pulsar/Twig/Extension/SanitiseExtensionTest.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->ext = new SanitiseExtension();
+    }
+
+    public function testGetName()
+    {
+        $this->assertEquals('sanitise_extension', $this->ext->getName());
```
[the remark this answers is not in the document]
Comment (Stanton) [reply tests/unit/Jadu/Pulsar/Twig/Extension/SanitiseExtensionTest.php:15]: Boilerplate from Twigs test examples

File: src/Jadu/Pulsar/Twig/Extension/SanitiseExtension.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        'cdata' => 0,
+        'deny_attribute' => 'on*',
+        'unique_ids' => 0,
+        'elements' => '* -applet -audio -canvas -embed -iframe -object -script -video',
+        'schemes' => 'href: aim, feed, file, ftp, gopher, http, https, irc, mailto, news, nntp, sftp, ssh, tel, telnet; style: !; *:file, http, https'
+    );
+
+    public function sanitise($data)
+    {
+        return Htmlawed::filter($data, self::FILTER_CONFIG);
```
[the remark this answers is not in the document]
Comment (Stanton) [reply src/Jadu/Pulsar/Twig/Extension/SanitiseExtension.php:33]: If there is a desire and request to override it then it can be added, otherwise it's premature complexity for a PR still open to discussion.

File: views/pulsar/v2/forms.html.twig [hunk 4]
```diff
@@ -191,7 +191,7 @@
                 {% if attr.help_text is defined and attr.help_text is not null %}
                     <p class="help-block" id="{{ helpId }}">{{ attr.help_text }}</p>
                 {% elseif attr.help_html is defined and attr.help_html is not null %}
-                    <p class="help-block" id="{{ helpId }}">{{ attr.help_html|raw }}</p>
+                    <p class="help-block" id="{{ helpId }}">{{ attr.help_html|sanitise|raw }}</p>
```
[the remark this answers is not in the document]
Comment (Stanton) [reply views/pulsar/v2/forms.html.twig:194]: `sanitise` wouldn't remove the HTML from the help text, it would still work as expected.

File: views/pulsar/components/tab.html.twig [hunk 5]
```diff
@@ -69,20 +69,20 @@
...
+            {{ _settings|sanitise|raw }}
         </section>
         {% endif %}
         <div class="tab__content" id="skip-target" role="main">
             {% block tab_content %}{% endblock %}
         </div>
         {% if _hasSidebar == true %}
         <aside class="tab__sidebar" aria-label="additional information">
-            {{ _sidebar|raw }}
+            {{ _sidebar|sanitise|raw }}
```
[the remark this answers is not in the document]
Comment (Stanton) [reply views/pulsar/components/tab.html.twig:85]: Agreed, as mentioned in the PR we probably don't want to sanitise structural/layout like this, and should focus on the helpers.

**Comment by Stanton** [comment]
This PR isn’t going into the software yet, it’s an option being tabled for discussion with POs and your concerns will definitely be considered, just please do view it all through the lens of ‘how hard will it be to get this change through CMS into XFP’. 

> If this PR were to be merged, it then becomes Pulsar who decides how data is sanitised and outputted. The control of the sanitisation of data is no longer the responsibility of the application using Pulsar, rather Pulsar itself. This is a bad thing.

If product owners are able to take that responsibility and sanitise data before it gets to the view, then I’m more than happy to work towards that goal. This is not currently the case.

The example in the PR of a user passing a malicious string which is rendered in a button helper is the basic example of Pulsar’s issue using `raw` combined with a lack of sanitisation on input by product teams. The button helper uses raw so that icons can also be rendered within the label, there’s no escape strategy for this, and if there is I’d love to know it.

The original solution was to re-work how and where we’re using raw, either making it opt-in, or reworking the helpers so that, for example, an icon can be passed as a helper option and not part of the label. That option is still on the table, however, it represents a much larger change which is fine as far as I’m concerned, but significantly more difficult to get approved for adoption by the product teams due to the amount of testing involved with that type of change in Pulsar.

> With this change, certain HTML in those "structural" variables would end up being stripped by this |sanitise filter.

Not really, entirely depends on the filtering configuration used, we can provide different filtering levels to control how strict the sanitisation is.

> Additionally, what this is achieving isn't what I'd expect when I pass user data to a helper. If I was passing a variable containing the user data <strong onclick='alert(2)'>foo</strong> then I would expect exactly that variable to be outputted as text. As it's user data, the HTML should not be respected at all and should be escaped entirely. And the value of the content should not change at all. This is crucial for things like form.content().

Back to buttons as a singular example, that’s not the way it works at the moment, buttons could be reworked to achieve this which would be the ideal solution from Pulsar’s point of view, but again, more difficult to get into CMS. Also count how many ‘icons in a button’ XFP uses.

Also places where you may use form.content to render a clickable email address, can all products can say hand-on-heart that a user attempting to provide a Little Bobby Tables email would be safe.

> Blindly adding a sanitise filter will cause no end of problems.

File: views/pulsar/v2/helpers/elem.html.twig (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackbentley) [reply views/pulsar/v2/helpers/elem.html.twig:423]: What about when `form__group--full` is applied? Does it get affected?

[the remark this answers is not in the document]
Comment (Stanton) [reply views/pulsar/v2/helpers/elem.html.twig:423]: They still work as expected

**Comment by jackbentley** [comment]
> If product owners are able to take that responsibility and sanitise data before it gets to the view, then I’m more than happy to work towards that goal. This is not currently the case.

I'd like to think this is the case currently. I do feel at least for the most part we take responsibility in XFP for such things. I obviously can't speak for other teams. The main issues we face is the inconsistency of when we do and don't need to do things like escaping/encoding.

> The example in the PR of a user passing a malicious string which is rendered in a button helper is the basic example of Pulsar’s issue using `raw` combined with a lack of sanitisation on input by product teams. The button helper uses raw so that icons can also be rendered within the label, there’s no escape strategy for this, and if there is I’d love to know it.

I think it's a bit of a moot point as I can't off the top of my head think of a good reason for user data to be outputted in a button. But if that ever is the case; providing you know that the text will be outputted using raw (this is the key part here - knowing so) then you can simply wrap the affected user data with the correct filter or use a set block. See https://twigfiddle.com/3wr5pw

> The original solution was to re-work how and where we’re using raw, either making it opt-in, or reworking the helpers so that, for example, an icon can be passed as a helper option and not part of the label. **That option is still on the table**, however, it represents a much larger change which is fine as far as I’m concerned, but significantly more difficult to get approved for adoption by the product teams due to the amount of testing involved with that type of change in Pulsar.

That would be my preferred approach so far. It might well be a larger changeset, but both solutions will end up being a breaking change. This option does however fix the root cause of the problem. That root problem being in the Pulsar API for the Twig helpers. This PR in its current state doesn't fix the root problem. It just fixes a symptom of the problem. And arguably causes other issues in the process. At the end of the day, this is a security issue and it needs to be fixed correctly.

> Back to buttons as a singular example, that’s not the way it works at the moment, buttons could be reworked to achieve this which would be the ideal solution from Pulsar’s point of view, but again, more difficult to get into CMS. Also count how many ‘icons in a button’ XFP uses. Changing the helpers is the easy bit, getting the change into the products is the hardest bit, this suggestion was a way of achieving the end goal by the view taking the responsibility of sanitisation.

Again, buttons are a bit of a bad example. They are unlikely to contain user data. There might be a couple of places that do, but still. In terms of getting the change into products, it's not a complicated logic change. It would be a hopefully simple API change. Hopefully with documentation on what to look for and what to change to - so mostly a find and replace exercise. I could imagine most the bulk of the work would be able to be achieved by regex find and replace if you're savvy enough with it.

As I mentioned though, regardless of which option we go down, it's likely to be a breaking change. Therefore the changes need to be well documented so that any updates in the products can we do swiftly with confidence.

> Also places where you may use form.content to render a clickable email address, can all products can say hand-on-heart that a user attempting to provide a Little Bobby Tables email would be safe?

This kind of falls into what I was saying about this being a product issue and not a Pulsar issue. If we know that the helper is going to be using `raw` and we pass HTML alongside a variable that isn't escaped, then it's the fault of the implementation - not Pulsar's. Again see https://twigfiddle.com/3wr5pw

> > Blindly adding a sanitise filter will cause no end of problems.
> 
> Perhaps you're right, and blindly rewriting helpers will also cause problems. Question is which approach will product teams prefer to go with?

They'll both cause problems if not documented well and are both "blindly applied". To get a good solution it needs to be carefully considered and engineered for all points that it affects.

As I've already mentioned, this PR is just fixing a symptom - it's like painting over a crack in a wall to hide it. I think the root issue here is that the API for receiving the content to output in the Pulsar helpers is not consistent nor documented well. In order to solve the root issue with the API, we would need to update the API to be consistent and document it well.

Without a better word to use, I think this option to solve the problem is just _whack_.

**Comment by Stanton** [comment]
#1340 has been selected in favour of sanitisation
